pipelines/smartseq2_single_sample/GroupQCOutputs.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Aug 21 11:07:46 2018
+
+@author: jishuxu
+"""
+
+from crimson import picard
+import argparse
+import os
+import pandas as pd
+
+def AggregatePicardMetricsRow(filenames, output_name):
+    """
+    pipeline output picard QC metrics at sinle cell/sample level.
+    This functin is called to merge/aggregate QC metrics by metrics type and
+    then merge multiple QC measurement into single matrix file.
+    In this file, column is sample/cell and row is QC metrics
+    :param files: metric files from pipeline outputs
+    :param met_name: metrics name with workflow name and subworkflow name
+    as prefix.such as 'run_pipelines.RunStarPipeline.alignment_summary_metrics'
+    """
+    # initial output
+    mets = {}
+    d = pd.DataFrame()
+    for file_name in filenames: 
+        cell_id = os.path.basename(file_name).split('_qc')[0]
+        mets[cell_id] = {}
+        parsed = picard.parse(file_name)
+        class_name = parsed['metrics']['class'].split('.')[2]
+        # Aignment metrics return multiple lines,
+        # but only output PAIRED-READS/third line
+        if class_name == "AlignmentSummaryMetrics":
+            # only parse out pair reads
+            met = parsed['metrics']['contents'][2]
+        # sometimes(very rare), insertion metrics also return multiple lines
+        # results to include TANDEM repeats. but we only output the first line.
+        elif class_name == "InsertSizeMetrics":
+            # if the elemnet counts is less than 21,
+            # it means insertion metrics returns multiple line results.
+            if len(parsed['metrics']['contents']) < 21:
+                met = parsed['metrics']['contents'][0]
+            else:
+                met = parsed['metrics']['contents']
+        else:
+            # other metrics(so far) only return one line results.
+            met = parsed['metrics']['contents']
+        mets[cell_id].update({
+                k: met[k] for k in met if k not in
+                ['SAMPLE', 'LIBRARY', 'READ_GROUP']
```

Comment:
I generally define any constant values in variables at the top of the file, and then reference those throughout the script